README.md
# gpustatd by m4rkw

gpustatd is a fan control daemon intended for use with Excavator by Nicehash. It
serves as a lightweight and decoupled mechanism for regulating fan speeds on
Nvidia cards when running intensive processes such as mining.


## Disclaimer

Please see the terms of the MIT LICENSE file included in this repo. gpustatd
does not guarantee that your card will stay cool under any conditions. If you
are using this tool to regulate temperatures under high load conditions you
should ensure that you have adequate monitoring

Although it is highly unlikely the author assumes no responsibility for any
damage whatsoever that may occur to your equipment when using this tool.

In testing with the default settings gpustatd consistently ran the fans
faster than the native Nvidia fan regulation.


## Usage

gpustatd monitors your Nvidia devices and regulates the fan speed according to
the card's temperature.


## Configuration

The configuration file should be installed to: /etc/gpustatd.conf

The config options are:

````
target_temperature: 75
fan_speed_temperature_ratio: 3
````

target_temperature is the temperature you want to keep the card at. At this
temperature the fan will run at 100%. The fan_speed_temperature_ratio option
controls how much gpustatd will decrease the fan speed by as the temperature
decreases. A ratio of 3 will drop the fan speed by 3% for every 1C below the
target_temperature.

````
temperature_limit: 80
````

If the card gets to this temperature, gpustatd will begin to throttle the
power limit to get the temperature down.

````
xorg_display_no: 0
````

Xorg is required to use nvidia-settings, so we need to know the display no.

````
automatically_reload_config_on_change: true
````

If this is enabled gpustatd will automatically reload its config when the
configuration file changes.

````
log_file: /var/log/gpustatd/gpustatd.log
logfile_max_size_mb: 100
logfile_max_count: 10
````

Logfile location and rotation config.


## Installation

Just install the RPM, or run through the following steps:

1. install PyYAML

````
sudo pip install PyYAML
````

2. edit gpustatd.conf.example as -> /etc/gpustatd.conf

3. ensure nvidia-settings has +s and that the user that gpustatd will run as has
permissions to use it

4. create /var/log/gpustatd and ensure the gpustatd user can write to it

5. simply run it:

````
./gpustatd.py
````

If you want to suppress the output to stdout, you can run gpustatd with:

````
$ ./gpustatd.py --silent
````

There is also a bash wrapper script to ensure that gpustatd is restarted if it
dies for any reason and a systemd unit so that gpustatd can be installed as a
system service.

If you notice any issues or crashes please file a bug on github.


## Related projects

- Minotaur - https://github.com/m4rkw/minotaur
- Excavataur - https://github.com/m4rkw/excavataur


## Donate

- XMR: 47zb4siDAi691nPW714et9gfgtoHMFnsqh3tKoaW7sKSbNPbv4wBkP11FT7bz5CwSSP1kmVPABNrsMe4Ci1F7Y2qLqT5ozd
- BTC: 1Bs4mCcyDcDCHfEisJqstEsmV5yzYcenJM


## Credits

- Much thanks to @gordan.bobic for packaging and invaluable input into the
  development of this tool


CHANGELOG.md
````
29/01/18 - 0.1 - initial release
29/01/18 - 0.2 - minor bugfix
30/01/18 - 0.3 - alert if fan speed == 0
01/02/18 - 0.4 - changes:
- added debug mode
- only allow one instance of fantotaur to be running
03/02/18 - 0.5 - changes:
- added "always monitor" config option
06/02/18 - 0.6 - changes:
- simplified and remove excavator API code
07/02/18 - 0.7 - changes:
- fixed auto config reload
12/02/18 - 0.8 - handle SIGINT/SIGTERM appropriately
12/02/18 - 0.8.1 - added fan speed floor
12/02/18 - 0.8.5 changes:
- replaced nvidia-smi with nvml
- indicate version on startup
15/02/18 - 0.9 changes:
- fanotaur is now gpustatd
18/02/18 - 1.0 changes:
- instead of restoring device fan control on exit, set fans to 85% and
set the default power limit
19/02/18 - 1.1 bugfix for setting appropriate power limits on startup
19/02/18 - 1.1.1: only set the power limit if its wrong
19/02/18 - 1.1.2: safety fix
19/02/18 - 1.1.3: switched to NVML
19/02/18 - 1.1.4: only invoke nvidia-settings once when setting fan speeds
````


log.py

import os
import datetime
import sys
from singleton import Singleton
from config import Config

class Log:
  __metaclass__ = Singleton

  def __init__(self):
    if "log_file" in Config().keys():
      self.log_file = Config().get('log_file')
    else:
      self.log_file = "/var/log/gpustatd/gpustatd.log"

    self.silent = False


  def silent(self):
    self.silent = True


  def add(self, level, message):
    if not os.path.exists(os.path.dirname(self.log_file)):
      os.mkdir(os.path.dirname(self.log_file))

    now = datetime.datetime.now()

    with open(self.log_file, "a+") as f:
      f.write("%s: [%s] %s\n" % (now.strftime("%Y-%m-%d %H:%M:%S"), level, message))

      if not self.silent:
        sys.stdout.write("%s: %s\n" % (level, message))
        sys.stdout.flush()

    if "logfile_max_size_mb" in Config().keys():
      max_size = Config().get('logfile_max_size_mb')
    else:
      max_size = 100

    if os.path.getsize(self.log_file) >= (max_size * 1024 * 1024):
      self.rotate_logs()

    if level == "fatal":
      sys.exit(1)


  def rotate_logs(self):
    logfile = self.log_file

    if "logfile_max_count" in Config().keys():
      max_count = Config().get('logfile_max_count')
    else:
      max_count = 10

    for i in reversed(range(1, max_count - 1)):
      os.rename(logfile + ".%d" % (i), logfile + ".%d" % (i+1))

    os.rename(logfile, logfile + ".1")


config.py

from singleton import Singleton
import os
import yaml
import sys

class Config(object):
  __metaclass__ = Singleton

  def load(self):
    self.config_file = "/etc/gpustatd.conf"

    if not os.path.exists(self.config_file):
      print "%s does not exist" % (self.config_file)
      sys.exit(1)

    self.config = yaml.load(open(self.config_file).read())
    self.config_mtime = os.stat(self.config_file).st_mtime
    self.ex_config_mtime = None

    for required in ["target_temperature", "fan_speed_temperature_ratio", "xorg_display_no"]:
      if not required in self.config.keys():
        print "required config key missing: %s" % (required)
        sys.exit(1)

  def reload(self):
    self.load()

  def reload_if_changed(self):
    if os.stat(self.config_file).st_mtime != self.config_mtime:
      print "info: config file change detected, reloading"
      self.reload()
      return True

    return False

  def get(self, key, data=None):
    if data == None:
      data = self.config

    if "." in key:
      segment = key.split('.')[0]

      if segment in data.keys():
        return self.get(key[len(segment)+1:], data[segment])

    if key not in data.keys():
      return None

    return data[key]

  def keys(self):
    return self.config.keys()


nvidia.py
#!/usr/bin/env python

import re
import os
import sys
import time
import yaml
import glob
import signal
from version import Version
from config import Config
from log import Log
from singleton import Singleton
from pynvml import *

class Nvidia(object):
  __metaclass__ = Singleton

  def __init__(self):
    os.environ["DISPLAY"] = ":%d" % (Config().get('xorg_display_no'))

    self.data_dir = "/var/run/gpustatd"
    self.devices = {}

    nvmlInit()

    self.handles = {}

    for line in os.popen("/usr/bin/nvidia-smi -L").read().rstrip().split("\n"):
      match = re.match("^GPU ([\d]+): (.*?) \(", line)
      if match:
        device_id = int(match.group(1))

        name = match.group(2)
        name = re.sub("^GeForce GTX ", "", name).replace(" ", "")

        handle = nvmlDeviceGetHandleByIndex(device_id)
        self.handles[device_id] = handle

        power_contraints = nvmlDeviceGetPowerManagementLimitConstraints(handle)

        default_power_limit = nvmlDeviceGetPowerManagementDefaultLimit(handle) / 1000

        self.devices[device_id] = {
          'id': device_id,
          'name': name,
          'has_fan': False,
          'min_power_limit_f': power_contraints[0] / 1000,
          'min_power_limit': '%d W' % (power_contraints[0] / 1000),
          'max_power_limit_f': power_contraints[1] / 1000,
          'max_power_limit': '%d W' % (power_contraints[1] / 1000),
          'default_power_limit_f': default_power_limit,
          'default_power_limit': '%d W' % (default_power_limit),
        }

    device_id = None

    for line in os.popen("/usr/bin/nvidia-settings -q fans 2>/dev/null").read().rstrip().split("\n"):
      match = re.match("^[\t\s]*\[([\d]+)\]", line)
      if match:
        device_id = int(match.group(1))
      if re.match("^[\s\t]*FAN-", line):
        self.devices[device_id]['has_fan'] = True


  def update_yaml(self, devices):
    device_ids = []

    for device_id in devices.keys():
      with open("%s/%d.yml.new" % (self.data_dir, devices[device_id]['id']), "w") as f:
        f.write(yaml.dump(devices[device_id]))
      os.rename("%s/%d.yml.new" % (self.data_dir, devices[device_id]['id']), "%s/%d.yml" % (self.data_dir, devices[device_id]['id']))

      device_ids.append(devices[device_id]['id'])

    for filename in glob.glob("%s/*.yml" % (self.data_dir)):
      device_id = filename.split("/")[-1].split(".")[0]

      if device_id.isdigit() and int(device_id) not in device_ids:
        os.remove(filename)


  def refresh(self, include_fans=False):
    for device_id in sorted(self.devices.keys()):
      util = nvmlDeviceGetUtilizationRates(self.handles[device_id])

      self.devices[device_id]['gpu_t_i'] = nvmlDeviceGetTemperature(self.handles[device_id], 0)
      self.devices[device_id]['gpu_t'] = '%d C' % (self.devices[device_id]['gpu_t_i'])
      self.devices[device_id]['power_f'] = float(nvmlDeviceGetPowerUsage(self.handles[device_id])) / 1000
      self.devices[device_id]['power'] = '%.2f W' % (self.devices[device_id]['power_f'])
      self.devices[device_id]['limit_f'] = nvmlDeviceGetEnforcedPowerLimit(self.handles[device_id]) / 1000
      self.devices[device_id]['limit'] = '%d W' % (self.devices[device_id]['limit_f'])
      self.devices[device_id]['gpu_u_i'] = util.gpu
      self.devices[device_id]['gpu_u'] = '%d %%' % (util.gpu)
      self.devices[device_id]['mem_u_i'] = util.memory
      self.devices[device_id]['mem_u'] = '%d %%' % (util.memory)
      self.devices[device_id]['gpu_f'] = '%d MHz' % (nvmlDeviceGetClockInfo(self.handles[device_id], 0))
      self.devices[device_id]['mem_f'] = '%d MHz' % (nvmlDeviceGetClockInfo(self.handles[device_id], 2))
      self.devices[device_id]['ps'] = 'P%d' % (nvmlDeviceGetPowerState(self.handles[device_id]))

    if include_fans:
      fan_speeds = self.get_all_fan_speeds()

      for device_id in self.devices.keys():
        if device_id in fan_speeds.keys():
          self.devices[device_id]['fan'] = fan_speeds[device_id]
        else:
          self.devices[device_id]['fan'] = None

    self.update_yaml(self.devices)

    return self.devices


  def get_all_fan_speeds(self):
    command = "/usr/bin/nvidia-settings"

    fan_map = {}
    i = 0

    for device_id in self.devices:
      if self.devices[device_id]['has_fan']:
        command += " -q '[fan:%d]/GPUTargetFanSpeed'" % (device_id)
        fan_map[i] = device_id
        i += 1

    if not Config().get('debug'):
      command += " 2>&1"

    resp = os.popen(command).read().rstrip()

    speeds = {}
    i = 0

    match = re.findall("Attribute.*?GPUTargetFanSpeed.*?\(.*?\):[\s\t\r\n]+([\d]+)", resp, re.MULTILINE)

    for m in match:
      speeds[fan_map[i]] = int(m)
      i += 1

    return speeds


  def set_fan_control_state(self, device_id, state):
    state = 1 if state else 0

    if not Config().get('debug'):
      suppress = '1>/dev/null 2>/dev/null'
    else:
      suppress = ''

    return os.system("/usr/bin/nvidia-settings -a '[gpu:%d]/GPUFanControlState=%d' %s" % (device_id, state, suppress)) == 0


  def get_fan_speed(self, device_id):
    return self.devices[device_id]['fan']


  def set_fan_speed(self, device_id, speed):
    if not Config().get('debug'):
      suppress = '1>/dev/null 2>/dev/null'
    else:
      suppress = ''

    if os.system("/usr/bin/nvidia-settings -a '[fan:%d]/GPUTargetFanSpeed=%d' %s" % (device_id, speed, suppress)) == 0:
      self.devices[device_id]['fan'] = speed
      self.update_yaml(self.devices)
      return True

    return False


  def apply_fan_speed_changes(self, speeds):
    cmd = "/usr/bin/nvidia-settings"

    for device_id in speeds.keys():
      cmd += " -a '[fan:%d]/GPUTargetFanSpeed=%d'" % (device_id, speeds[device_id])

    if not Config().get('debug'):
      cmd += " 1>/dev/null 2>/dev/null"

    if os.system(cmd) == 0:
      for device_id in speeds.keys():
        self.devices[device_id]['fan'] = speeds[device_id]
      self.update_yaml(self.devices)
      return True

    return False


  def set_all_fans(self, speed):
    cmd = "/usr/bin/nvidia-settings"

    for device_id in self.devices.keys():
      cmd += " -a '[fan:%d]/GPUTargetFanSpeed=%d'" % (device_id, speed)

    if not Config().get('debug'):
      cmd += " 1>/dev/null 2>/dev/null"

    if os.system(cmd) == 0:
      for device_id in self.devices.keys():
        self.devices[device_id]['fan'] = speed
      self.update_yaml(self.devices)
      return True

    return False



  def set_power_limit(self, device_id, watts):
    return os.system("/usr/bin/nvidia-smi -i %d --power-limit=%d 1>/dev/null 2>/dev/null" % (device_id, watts)) == 0


gpustatd.py
#!/usr/bin/env python

from nvidia import Nvidia
from config import Config
import time
from log import Log
import os
import signal
import sys
from pynvml import *
from version import Version

class GPUStatD:
  def __init__(self):
    Config().load()

    if len(sys.argv) >1 and sys.argv[1] == "--silent":
      Log().silent()

    Log().add('info', 'gpustatd %s starting up' % (Version().get()))

    if self.already_running():
      Log().add('fatal', 'gpustatd is already running')

    self.create_pid_file()
    self.throttled = []

    self.target_temperature = Config().get('target_temperature')
    self.fan_speed_temperature_ratio = Config().get('fan_speed_temperature_ratio')
    os.environ["DISPLAY"] = ":%d" % (Config().get('xorg_display_no'))

    Log().add('info', 'scanning devices')
    self.devices = Nvidia().refresh(True)
    self.speeds = {}

    for device_id in self.devices.keys():
      self.speeds[device_id] = self.devices[device_id]['fan']

    self.self_test()


  def already_running(self):
    if not os.path.exists("/var/run/gpustatd"):
      try:
        ok.mkdir("/var/run/gpustatd", 0755)
      except:
        Log().add('fatal', 'unable to create /var/run/gpustatd')

    if os.path.exists("/var/run/gpustatd/gpustatd.pid"):
      pid = open("/var/run/gpustatd/gpustatd.pid").read().rstrip()

      return pid.isdigit() and os.path.exists("/proc/%s" % (pid))


  def create_pid_file(self):
    if self.already_running():
      Log().add('fatal', 'unable to start, there is another gpustatd process running')

    with open("/var/run/gpustatd/gpustatd.pid", "w") as f:
      f.write("%d" % (os.getpid()))


  def self_test(self):
    self.fan_states = {}

    Log().add('info', 'self-test')

    self.do_self_test()

    Log().add('info', 'looks good, monitoring')

    signal.signal(signal.SIGINT, self.sigint_handler)
    signal.signal(signal.SIGTERM, self.sigint_handler)
    signal.signal(signal.SIGHUP, self.sighup_handler)

    self.start()


  def sighup_handler(self):
    Log().add('info', 'HUP received, reloading config')

    Config().load()


  def sigint_handler(self, a, b):
    Log().add('info', 'interrupt received, setting fans to 85% and cards to default power limit')

    Nvidia().set_all_fans(85)

    for device_id in self.devices.keys():
      if self.devices[device_id]['default_power_limit_f']:
        Nvidia().set_power_limit(device_id, self.devices[device_id]['default_power_limit_f'])

    sys.exit(0)


  def do_self_test(self):
    for device_id in self.devices.keys():
      device = self.devices[device_id]
      if not "ignore_devices" in Config().keys() or not Config().get('ignore_devices') or device['id'] not in Config().get('ignore_devices'):
        if not Nvidia().set_fan_control_state(device['id'], 1):
          Log().add('fatal', 'failed to toggle the fan control state for device: %d' % (device['id']))
        self.fan_states[device['id']] = True
        self.regulate_power_level(device, True)

    return True


  def start(self):
    i = 0

    while True:
      if i % 10 == 0:
        self.devices = Nvidia().refresh(True)
      else:
        self.devices = Nvidia().refresh()

      for device_id in self.devices.keys():
        device = self.devices[device_id]
        if "ignore_devices" in Config().keys() and Config().get('ignore_devices') and device['id'] in Config().get('ignore_devices'):
          continue

        self.speeds[device_id] = self.calculate_fan_speed(device)
        self.regulate_power_level(device)

      self.adjust_fan_speeds()

      if "automatically_reload_config_on_change" in Config().keys() and Config().get("automatically_reload_config_on_change"):
        if Config().reload_if_changed():
          self.target_temperature = Config().get('target_temperature')
          self.fan_speed_temperature_ratio = Config().get('fan_speed_temperature_ratio')
          os.environ["DISPLAY"] = ":%d" % (Config().get('xorg_display_no'))

      time.sleep(1)
      i += 1


  def ensure_control_state(self, device_id, state):
    if self.fan_states[device_id] != state:
      Nvidia().set_fan_control_state(device_id, state)
      self.fan_states[device_id] = state


  def calculate_fan_speed(self, device):
    if device['gpu_t_i'] >= self.target_temperature:
      fan_speed = 100
    else:
      fan_speed = 100 - ((self.target_temperature - device['gpu_t_i']) * self.fan_speed_temperature_ratio)

    if fan_speed <0:
      fan_speed = 20

    return fan_speed


  def adjust_fan_speeds(self):
    fan_speed_changes = {}

    for device_id in self.devices.keys():
      if self.devices[device_id]['fan'] != self.speeds[device_id]:
        if Config().get('informative'):
          Log().add('debug', '%d: device temp: %s  target temperature: %s  target fan speed: %d' % (device_id, self.devices[device_id]['gpu_t_i'], self.target_temperature, self.speeds[device_id]))

        fan_speed_changes[device_id] = self.speeds[device_id]

    if len(fan_speed_changes) >0:
      Nvidia().apply_fan_speed_changes(fan_speed_changes)


  def regulate_power_level(self, device, force=False):
    card_limit = Config().get("temperature_limit")

    if not card_limit:
      card_limit = 80

    if os.path.exists("/var/run/minotaur/%d.powerlimit" % (device['id'])):
      desired_limit = float(open("/var/run/minotaur/%d.powerlimit" % (device['id'])).read())
    else:
      desired_limit = device['default_power_limit_f']

    temp = device['gpu_t_i']
    limit = device['limit_f']

    if temp >= card_limit:
      deduction = 10 * (2 ** (temp - card_limit))

      new_limit = desired_limit - deduction

      if new_limit < device['min_power_limit_f']:
        new_limit = device['min_power_limit_f']

      if new_limit != limit:
        Log().add('warning', 'device %d: temperature is at %dC, throttling power limit to %dW' % (device['id'], temp, new_limit))
        Nvidia().set_power_limit(device['id'], new_limit)

        if not device['id'] in self.throttled:
          self.throttled.append(device['id'])
    else:
      if device['id'] in self.throttled and limit < desired_limit:
        Log().add('info', 'device %d: temperature is at %dC, restoring optimum power limit of %dW' % (device['id'], temp, desired_limit))

        Nvidia().set_power_limit(device['id'], desired_limit)

        self.throttled.remove(device['id'])
      elif force:
        if int(device['limit_f']) != int(desired_limit):
          Log().add('info', 'device %d: setting power limit %dW' % (device['id'], desired_limit))
          Nvidia().set_power_limit(device['id'], desired_limit)

GPUStatD()
